src/wasm/wasm-s-parser.cpp

```diff
@@ -925,10 +925,34 @@ Type SExpressionWasmBuilder::elementToType(Element& s) {
   if (s.isStr()) {
     return stringToType(s.str(), false, false);
   }
-  auto& tuple = s.list();
+  auto& list = s.list();
+  auto size = list.size();
+  if (size > 0 && elementStartsWith(s, REF)) {
+    // It's a reference. It should be in the form
+    //   (ref $name)
+    // or
+    //   (ref null $name)
+    if ((size != 2 && size != 3) || !list[1]->isStr()) {
+      throw ParseException(
+        std::string("invalid reference type"), s.line, s.col);
+    }
```

Comment:
While we are checking, how about checking both ways, and all at the top, like
```suggestion
    if (size != 2 && size != 3) {
      throw ParseException(
        std::string("invalid reference type"), s.line, s.col);
    }
    if (size == 2 && !list[1]->isStr()) {
      throw ParseException(
        std::string("invalid reference type"), s.line, s.col);
    }
    if (size == 3 && (*list[1] != NULL_ || !list[2]->isStr())) {
      throw ParseException(
        std::string("invalid reference type"), s.line, s.col);
}
```